### DISCUSSION TITLE
Difference between p2 folders in local repository

### DISCUSSION THREAD

**Discussion by lppedd** [eclipse-tycho/tycho#2037] [Q&A]
In the local repository, we can find a `p2` folder.
In my case I have the following layout:

![image](https://user-images.githubusercontent.com/19871649/215498265-1f9fbf87-95ee-4796-b884-054b6e12446b.png)

What are the differences between these folders? How are they used?

---

Continuing the https://github.com/eclipse-tycho/tycho/discussions/2031 discussion on resolving dependencies.
That works for plugins, but not for features.

When manually resolving a _feature_ artifact throught `WorkspaceReader`s, `TychoWorkspaceReader` composes the local path as, for example:
```
C:\Users\edoardo.luppi\.m2\repository\p2\osgi\bundle\org.eclipse.nebula.widgets.nattable.core.feature\2.0.0.202012161455\org.eclipse.nebula.widgets.nattable.core.feature-2.0.0.202012161455.jar
```

But that folder only contains plugins. Features seems to be located starting directly at the root of the `p2` folder:
```
C:\Users\edoardo.luppi\.m2\repository\p2\org\eclipse\update\feature\org.eclipse.nebula.widgets.nattable.core.feature\2.0.0.202012161455\org.eclipse.nebula.widgets.nattable.core.feature-2.0.0.202012161455.jar
```

So the question is, how do I correctly resolve/download feature JARs?

**Comment by laeubi** [comment]
> So the question is, how do I correctly resolve/download feature JARs?

Feature jars are usually not something that can participate in maven dependency resolution, so you should explain what "resolve" means and what "do not work" mean.

**Reply by lppedd** [reply]
Ok. I'll rephrase after investigating more.

First of all, let's say we want to resolve (in case it's not in the local repo, we want to fetch it) the following plugin artifact:

![image](https://user-images.githubusercontent.com/19871649/215508954-82cbc3ac-fa75-4365-84d5-b3482d31f0d0.png)

That artifact has been extracted via `project.getDependencyArtifacts()`.

Using `session.getRequest().getWorkspaceReader().findArtifact(artifact)` is going to work **only** if the artifact is already downloaded under `p2/osgi/bundle/...`. See below snippet.

https://github.com/eclipse-tycho/tycho/blob/66e864fecb7a3aa78c4b4e34749d3aecb491115e/tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoWorkspaceReader.java#L95-L98

If `cachedFile.isFile()` returns `false`, the artifact key is compared with a `MavenDependencyDescriptor dependencyDescriptor`.  
But look at how that object differs from the input artifact:

![image](https://user-images.githubusercontent.com/19871649/215511559-ac791ba3-c4f3-4926-b021-ea2d038f56fc.png)

Then, the two objects get compared by `groupId` and `artifactId`.

https://github.com/eclipse-tycho/tycho/blob/66e864fecb7a3aa78c4b4e34749d3aecb491115e/tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoWorkspaceReader.java#L112-L113

You can see the comparison with the `groupId` is always going to fail.  
So the artifact is never fetched with `descriptor.getLocation(true)`.

**Reply by lppedd** [reply]
This seems like a bug to me. It's expected the `artifact.groupId()` to begin with `p2`

https://github.com/eclipse-tycho/tycho/blob/66e864fecb7a3aa78c4b4e34749d3aecb491115e/tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoWorkspaceReader.java#L93

And then the same `p2.*` `groupId` gets compared with a Maven `groupId`.

**Reply by laeubi** [reply]
> It's expected the artifact.groupId() to begin with p2

All "real" p2 artifacts begin with "p2." by definition, but many P2 sites promote bogus meta-data (see https://mvnrepository.com/artifact/org.eclipse.birt/org.eclipse.birt.core.ui that there is no 4.9.0-SNAPSHOT version on central).

So I think here is a check missing that the repository flag is set before augment the maven data (see `org.eclipse.tycho.core.maven.MavenDependencyInjector.isValidMavenDescriptor(MavenDependencyDescriptor)`) .

> This seems like a bug to me. 

As there is no feature / test associated with your use-case this is maybe a missing feature but not a bug ;-)
But if you like we can enhance this, just test this carefully to watch out for side-effects due to additional downloads.

**Reply by lppedd** [reply]
@laeubi 

> But if you like we can enhance this

That would be awesome.  
I can see that `MavenDependencyInjector` already uses validation

https://github.com/eclipse-tycho/tycho/blob/190dd00d45bb2b180d13ec76acf3d77a8febbdf1/tycho-core/src/main/java/org/eclipse/tycho/core/maven/MavenDependencyInjector.java#L282-L287

And in fact the BIRT group id is set to `p2.*`.

So probably it's a matter of saying "are these Maven coordinates ok? If not, let's fall back to the standard ones" in `TychoWorkspaceReader`.

**Reply by laeubi** [reply]
> And in fact the BIRT group id is set to p2.*.

This does not matter much as likely no one will ever deploy such an artifact, I think the `MavenDependencyDescriptor` is not useful in this code path but belong to the one marked with `TODO` below... I just experimented a lot with that but finally restricted the lookup to "plain" P2 items so this seems to slipped through.

So to summarize:

- if it starts with p2 `MavenDependencyDescriptor` should not be used
- if not, `MavenDependencyDescriptor` might be used to discover "secret" p2 artifacts, but we should disable that for now.

**Reply by lppedd** [reply]
@laeubi I can open a PR targeting `TychoWorkspaceReader` if you want.

> but we should disable that for now

You mean that this behavior of looking for Maven coordinates should be disabled/not present everywhere or just for this use case?

**Reply by laeubi** [reply]
The looking for maven coordinates should be moved to the else but commented out for now (as it is effectively inactive at the moment anyways).
